### PR TITLE
KEP-3325: Promote SelfSubjectReview to GA

### PR DESCRIPTION
**Pull request by nabokihms**
#### What type of PR is this?

/kind feature
/kind api-change

#### Which issue(s) this PR fixes:

This PR is a part of https://github.com/kubernetes/enhancements/issues/3325

#### Does this PR introduce a user-facing change?

```release-note
The `SelfSubjectReview` API is promoted to `authentication.k8s.io/v1` and the `kubectl auth whoami` command is GA.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

**Review by liggitt — COMMENTED** [review]
lgtm with a couple tweaks

File: pkg/features/kube_features.go [hunk 1]
```diff
@@ -879,7 +880,7 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 
 	AnyVolumeDataSource: {Default: true, PreRelease: featuregate.Beta}, // on by default in 1.24
 
-	APISelfSubjectReview: {Default: true, PreRelease: featuregate.Beta}, // on by default in 1.27
+	APISelfSubjectReview: {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // GA in 1.28
```
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/features/kube_features.go:883]: no, drop the feature-gate, sorry

File: staging/src/k8s.io/api/authentication/v1/types.go [hunk 2]
```diff
@@ -197,3 +197,29 @@ type BoundObjectReference struct {
 	// +optional
 	UID types.UID `json:"uid,omitempty" protobuf:"bytes,4,opt,name=uID,casttype=k8s.io/apimachinery/pkg/types.UID"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +genclient:onlyVerbs=create
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:prerelease-lifecycle-gen:introduced=1.28
```
[the remark this answers is not in the document]
Comment (nabokihms) [reply staging/src/k8s.io/api/authentication/v1/types.go]: Removed :+1:

File: pkg/features/kube_features.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nabokihms) [reply pkg/features/kube_features.go]: Added

**Comment by liggitt** [comment]
since this is promoting an API to v1 without requiring it to be in conformance, this should add to `test/conformance/testdata/ineligible_endpoints.yaml`

```yaml
- endpoint: createAuthenticationV1SelfSubjectReview
  reason: Cluster providers are allowed to choose to not serve this API, and the whoami command handles unavailability gracefully.
  link: https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/3325-self-subject-attributes-review-api/README.md#ga
```

File: test/e2e/auth/selfsubjectreviews.go [hunk 3]
```diff
@@ -22,6 +22,7 @@ import (
 
 	"github.com/onsi/ginkgo/v2"
 	"github.com/onsi/gomega"
+	authenticationv1 "k8s.io/api/authentication/v1"
 	authenticationv1alpha1 "k8s.io/api/authentication/v1alpha1"
 	authenticationv1beta1 "k8s.io/api/authentication/v1beta1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (liggitt) [inline test/e2e/auth/selfsubjectreviews.go:28]: the e2e test should drop the `[Feature:APISelfSubjectReview]` tag so it actually runs in the main e2e job (and may need to make the alpha/beta portions conditional on availability of those APIs)

File: test/cmd/auth_whoami.sh [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+# Runs tests for kubectl diff
+run_kubectl_events_tests() {
```
Comment (liggitt) [inline test/cmd/auth_whoami.sh:22]: typo/duplicate? rename to `run_kubectl_whoami_tests`?

Comment (liggitt) [reply test/cmd/auth_whoami.sh:22]: and add invocation of `run_kubectl_whoami_tests` from `test/cmd/legacy-script.sh`

File: test/e2e/auth/selfsubjectreviews.go (hunk 3 above)
Comment (nabokihms) [reply test/e2e/auth/selfsubjectreviews.go:29]: To not forget: also required to delete it from the list here https://github.com/kubernetes/test-infra/blob/a66784ee843eeaa97a7a13b756301d1e9a2c91ab/config/jobs/kubernetes/sig-cloud-provider/gcp/gcp-gce.yaml#L363

File: test/cmd/auth_whoami.sh (hunk 4 above)
Comment (nabokihms) [reply test/cmd/auth_whoami.sh]: My bad, fixed this

**Comment by liggitt** [comment]
/lgtm
/approve

/assign @dims 
for conformance file touch

**Pull request merged**